distributed/worker_state_machine.py

```diff
@@ -3602,6 +3583,77 @@ def _handle_stimulus_from_task(self, task: asyncio.Task[StateMachineEvent]) -> N
             return
         self.handle_stimulus(stim)
 
+    def _finalize_metrics(
+        self,
+        stim: StateMachineEvent,
+        ledger: DelayedMetricsLedger,
+    ) -> None:
+        activity: tuple[str, ...]
+        coarse_time: str | Literal[False]
+
+        if isinstance(stim, GatherDepSuccessEvent):
+            activity = ("gather-dep",)
+            for key in stim.data:
+                ts = self.state.tasks.get(key)
+                if ts and ts.state == "memory":
+                    # At least one key was fetched and stored in memory
+                    coarse_time = False
+                    break
+            else:
+                coarse_time = "cancelled" if stim.data else "missing"
```

Comment:
This logic is probably fragile in case of edge cases.
Luckily, we do not care about correctly categorizing the metric in case of a once-in-a-bluemoon use case :)